### PR TITLE
[Proof] Prevent proof submission when not required

### PR DESCRIPTION
**Pull request by red-0ne**
## Summary

Prevent the `SubmitProof` keeper from upserting a proof if the `Proof` is not required.

## Issue

`Proof`s are among the most block space-consuming primitives in Poktroll. 

In some cases, a `Proof` submission is unnecessary, but the current `SubmitProof` keeper still allows non-required proofs to be submitted.

The `SubmitProof` keeper has enough information to determine whether the `Proof` corresponding to a claim is necessary. The protocol could leverage this information to avoid saving non-required `Proof`s on-chain, thereby optimizing block space usage.

## Type of change

Select one or more from the following:

- [x] New feature, functionality or library
- [ ] Consensus breaking; add the `consensus-breaking` label if so. See #791 for details
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

- [ ] **Documentation**: `make docusaurus_start`; only needed if you make doc changes
- [x] **Unit Tests**: `make go_develop_and_test`
- [x] **LocalNet E2E Tests**: `make test_e2e`
- [ ] **DevNet E2E Tests**: Add the `devnet-test-e2e` label to the PR.

## Sanity Checklist

- [x] I have tested my changes using the available tooling
- [x] I have commented my code
- [x] I have performed a self-review of my own code; both comments & source code
- [ ] I create and reference any new tickets, if applicable
- [ ] I have left TODOs throughout the codebase, if applicable

### CODE REVIEW

File: x/proof/keeper/msg_server_submit_proof.go [hunk 1]
```diff
@@ -191,3 +201,110 @@ func (k Keeper) deductProofSubmissionFee(ctx context.Context, supplierOperatorAd
 
 	return nil
 }
+
+// proofRequirementForClaim checks if a proof is required for a claim.
+// If it is not, the claim will be settled without a proof.
+// If it is, the claim will only be settled if a valid proof is available.
+// TODO_BLOCKER(@bryanchriswhite, #419): Document safety assumptions of the probabilistic proofs mechanism.
+func (k Keeper) ProofRequirementForClaim(ctx context.Context, claim *types.Claim) (_ types.ProofRequirementReason, err error) {
```
Comment (Olshansk) [inline x/proof/keeper/msg_server_submit_proof.go:209]: Did you make changes here or just copy-paste? Haven't re-reviewed it yet.

**Review by Olshansk — APPROVED** [review]
Couple small things but pretty straight forward!

File: x/proof/keeper/proof_requirement_test.go [hunk 2]
```diff
@@ -4,26 +4,25 @@ import (
...
-	prooftypes "github.com/pokt-network/poktroll/x/proof/types"
+	"github.com/pokt-network/poktroll/x/proof/types"
 )
 
 func TestKeeper_IsProofRequired(t *testing.T) {
 	// Set expectedCompute units to be below the proof requirement threshold to only
 	// exercise the probabilistic branch of the #isProofRequired() logic.
-	expectedComputeUnits := prooftypes.DefaultProofRequirementThreshold - 1
-	keepers, ctx := keeper.NewTokenomicsModuleKeepers(t, log.NewNopLogger())
+	expectedComputeUnits := types.DefaultProofRequirementThreshold - 1
```
[the remark this answers is not in the document]
Comment (red-0ne) [reply x/proof/keeper/proof_requirement_test.go:20]: Yes, it's in #795, Should be fixed when merging.

File: x/proof/keeper/msg_server_submit_proof.go (hunk 1 above)
Comment (red-0ne) [reply x/proof/keeper/msg_server_submit_proof.go:209]: Just copy pasting, imports got adjusted though.

File: x/proof/keeper/msg_server_submit_proof.go [hunk 3]
```diff
@@ -191,3 +202,110 @@ func (k Keeper) deductProofSubmissionFee(ctx context.Context, supplierOperatorAd
 
 	return nil
 }
+
+// proofRequirementForClaim checks if a proof is required for a claim.
```
Comment (Olshansk) [inline x/proof/keeper/msg_server_submit_proof.go:206]: Does this even need to be exposed?

```suggestion
// ProofRequirementForClaim checks if a proof is required for a claim.
```

File: x/proof/keeper/msg_server_submit_proof.go [hunk 4]
```diff
@@ -191,3 +202,110 @@ func (k Keeper) deductProofSubmissionFee(ctx context.Context, supplierOperatorAd
...
+	numClaimComputeUnits, err = claim.GetNumComputeUnits()
+	if err != nil {
+		return requirementReason, err
+	}
+
+	proofParams := k.GetParams(ctx)
+
+	// Require a proof if the claim's compute units meets or exceeds the threshold.
+	//
+	// TODO_BLOCKER(@bryanchriswhite, #419): This is just VERY BASIC placeholder logic to have something
```
Comment (Olshansk) [inline x/proof/keeper/msg_server_submit_proof.go:232]:
```suggestion
	// TODO_BLOCKER(@Olshansk, #419): This is just VERY BASIC placeholder logic to have something
```

File: x/proof/keeper/msg_server_submit_proof.go [hunk 5]
```diff
@@ -191,3 +202,110 @@ func (k Keeper) deductProofSubmissionFee(ctx context.Context, supplierOperatorAd
...
+
+		logger.Info(fmt.Sprintf(
+			"claim requires proof due to compute units (%d) exceeding threshold (%d)",
+			numClaimComputeUnits,
+			proofParams.GetProofRequirementThreshold(),
+		))
+		return requirementReason, nil
+	}
+
+	earliestProofCommitBlockHash, err := k.getEarliestSupplierProofCommitBlockHash(ctx, claim)
```
Comment (Olshansk) [inline x/proof/keeper/msg_server_submit_proof.go:250]:
```suggestion
	// Hash of block when proof submission is allowed
	earliestProofCommitBlockHash, err := k.getEarliestSupplierProofCommitBlockHash(ctx, claim)
```

Comment (Olshansk) [inline x/proof/keeper/msg_server_submit_proof.go:250]: @red-0ne Can you double check that we reject proofs that are too early please>

File: x/proof/keeper/msg_server_submit_proof.go [hunk 6]
```diff
@@ -191,3 +202,110 @@ func (k Keeper) deductProofSubmissionFee(ctx context.Context, supplierOperatorAd
...
+	if err != nil {
+		return requirementReason, err
+	}
+
+	proofRequirementSampleValue, err := claim.GetProofRequirementSampleValue(earliestProofCommitBlockHash)
```
Comment (Olshansk) [inline x/proof/keeper/msg_server_submit_proof.go:255]:
```suggestion
	// The probability that a proof is required.
	proofRequirementSampleValue, err := claim.GetProofRequirementSampleValue(earliestProofCommitBlockHash)
```

File: x/proof/keeper/msg_server_submit_proof.go [hunk 7]
```diff
@@ -191,3 +202,110 @@ func (k Keeper) deductProofSubmissionFee(ctx context.Context, supplierOperatorAd
...
+		numClaimComputeUnits,
+		proofParams.GetProofRequirementThreshold(),
+		proofRequirementSampleValue,
+		proofParams.GetProofRequestProbability(),
+	))
+	return requirementReason, nil
+}
+
+// getEarliestSupplierProofCommitBlockHash returns the block hash of the earliest
+// block at which a claim might have its proof committed.
```
Comment (Olshansk) [inline x/proof/keeper/msg_server_submit_proof.go:285]:
```suggestion
// block at which a claim may have its proof committed.
```

**Review by Olshansk — APPROVED** [review]

File: x/proof/keeper/msg_server_submit_proof.go (hunk 3 above)
Comment (red-0ne) [reply x/proof/keeper/msg_server_submit_proof.go:206]: Yes, it's used in both `proof` and `tokenomics` keepers

File: x/proof/keeper/msg_server_submit_proof.go (hunk 5 above)
Comment (red-0ne) [reply x/proof/keeper/msg_server_submit_proof.go:250]: Yes it is, `k.validateProofWindow` is called before `k.ProofRequirementForClaim` in `k.SubmitProof`

**Pull request merged**